truffe2/main/test_tools.py

```diff
@@ -0,0 +1,179 @@
+# -*- coding: utf-8 -*-
+'''
+Unit test tools for truffe
+
+'''
+from bs4 import BeautifulSoup
+from json import loads
+import os
+import sys
+from _io import StringIO
+from os.path import dirname
+
+from django.test import TestCase, Client
+from django.test.client import MULTIPART_CONTENT
+from django.core.management import call_command
+
+from main.test_data import initial_data
+
+
+class TruffeTestAbstract(TestCase, Client):
+
+    def __init__(self, methodName):
+        TestCase.__init__(self, methodName)
+        Client.__init__(self)
+        self.send_content_type = MULTIPART_CONTENT
+
+    def setUp(self):
+        TestCase.setUp(self)
+        initial_data()
+        class_name = self.__class__.__name__
+        if 'WithLogin' in class_name:
+            self.connect_to('admin')
+        elif 'WithLog' in class_name:
+            user_num = int(class_name[class_name.index('WithLog') + 7])
+            self.connect_to('user%d' % user_num)
+
+    def connect_to(self, username):
+        self.call_check_html('/users/login')
+        csr_token_input = self.content.findAll('input', {'name': 'csrfmiddlewaretoken'})[0]           
+        self.call_check_redirect('/users/login', {'username': username, 'password': username, 'csrfmiddlewaretoken': csr_token_input}, 'post', '/')
+        self.call_check_html('/')
+
+    def _get_default_content_type(self):
+        if self.response.status_code != 200:
+            return ""
+        if '<!DOCTYPE html>' in self.response.content:
+            default_type = 'text/html'
+        elif len(self.response.content) == 0:
+            default_type = 'text/text'
+        elif self.response.content.strip()[0] in ('[', '{'):
+            default_type = 'application/json'
+        elif self.response.content[1:4] == 'PDF':
+            default_type = 'application/pdf'
+        else:
+            default_type = 'text/text'
+        return default_type
+
+    def call(self, path, data={}, method='get', status_expected=200):
+        self.response = getattr(self, method.lower(), self.get)(path, data, content_type=self.send_content_type)
+        self.content_type = getattr(self.response, 'content_type', self._get_default_content_type())
+        self.content = self.response.content
+        if status_expected != 0:
+            self.assertEqual(self.response.status_code, status_expected, "HTTP error [%s]:%s" % (self.response.status_code, self.get_div('text-center error-box')))
+
+    def call_check_redirect(self, path, data={}, method='get', redirect_url=None, target_status_code=200):
```

Comment:
This test essentially tests whether a view is decorated with `@login_required`. Is it efficient to test all of theses login protected functions by checking the redirect, or could you test for the presence of the decorator?

Reply:
It's juste a controle that no "login_required" is missing
And i have identified that "/accounting/main/accounting/import/step/0" has no decorator.
So, it's a no-regression control: check that nobody don't remove decorator in futur